### PR TITLE
UPPSF-3196 - minor/UPPSF-3196-update-clientgo

### PR DESCRIPTION
**Pull request by dtvalk-ov**
Update k8s' client-go to v0.24.0
Fix lib calls and tests
Bump go version to 1.18

## Why

[Jira Ticket](https://financialtimes.atlassian.net/browse/UPPSF-3196)

## Anything, in particular, you'd like to highlight to reviewers

Mention here sections of code which you would like reviewers to pay extra attention to .E.g

_Would appreciate a second pair of eyes on the test_  
_I am not quite sure how this bit works_  
_Is there a better library for doing x_  

## Scope and particulars of this PR (Please tick all that apply)

- [x] Tech hygiene (dependency updating & other tech debt)
- [ ] Bug fix
- [ ] Feature
- [ ] Documentation
- [ ] Breaking change
- [x] Minor change (e.g. fixing a typo, adding config)

___
This Pull Request follows the rules described in our [Pull Requests Guide](https://github.com/Financial-Times/upp-docs/tree/master/guides/pr-guide)

### CODE REVIEW

File: controller.go [hunk 1]
```diff
@@ -144,20 +145,22 @@ func (c *healthCheckController) runServiceChecksByServiceNames(services map[stri
 	}).Checks
 
 	wg := sync.WaitGroup{}
+	tempCtx, cancel := context.WithCancel(context.Background())
```
Comment (atanasdinov) [inline controller.go:148]: Is there a benefit of creating a temporary context over using the provided request one?

**Review by atanasdinov — COMMENTED** [review]
LGTM! Just wondering about the decisions behind selecting the specific contexts in the scenarios below.

File: cachingController.go [hunk 2]
```diff
@@ -94,14 +95,17 @@ func (c *healthCheckController) scheduleCheck(mService measuredService, refreshP
 		return
 	}
 
+	ctx, cancel := context.WithCancel(context.Background())
```
[the remark this answers is not in the document]
Comment (dtvalk-ov) [reply cachingController.go:98]: As this is a recursive function (called in a goroutine), I was concerned about creating a new context over and over each time without canceling it explicitly. The cancel is deferred right under the create statement

File: controller.go (hunk 1 above)
Comment (dtvalk-ov) [reply controller.go:148]: The original context gets canceled midway through the iteration and it causes errors:
`cannot build health check result for services: failed to get the categories from kubernetes: client rate limiter Wait returned an error: context canceled `
That's why I've used this workaround

**Review by atanasdinov — APPROVED** [review]

**Review by MiroslavGatsanoga — APPROVED** [review]

**Pull request merged**